### PR TITLE
IBD using chainwork instead of height and not using header timestamps

### PR DESCRIPTION
**Pull request by gmaxwell**
This introduces a 'minimum chain work' chainparam which is intended
 to be the known amount of work in the chain for the network at the
 time of software release.  If you don't have this much work, you're
 not yet caught up.
    
This is used instead of the count of blocks test from checkpoints.

This criteria is trivial to keep updated as there is no element of
 subjectivity, trust, or position dependence to it. It is also a more
 reliable metric of sync status than a block count.

This is the first step in a larger change to eliminate checkpoints
 completely, but it's independently a good idea.

This PR makes IsInitialBlockDownload no longer use header-only timestamps.

This avoids a corner case (mostly visible on testnet) where bogus
 headers can keep nodes in IsInitialBlockDownload.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
utACK f3f8d2e

Nit: `GetTotalBlocksEstimate` is no longer used (only in tests). Feel free to remove that now.

**Review by dcousens — APPROVED** [review]

**Comment by Leviathn** [comment]
utACK - certainly a more reliable metric.

**Comment by gmaxwell** [comment]
@MarcoFalke nits addressed, I think.

File: src/chainparams.cpp [hunk 1]
```diff
@@ -96,6 +96,9 @@ class CMainParams : public CChainParams {
         consensus.vDeployments[Consensus::DEPLOYMENT_SEGWIT].nStartTime = 1479168000; // November 15th, 2016.
         consensus.vDeployments[Consensus::DEPLOYMENT_SEGWIT].nTimeout = 1510704000; // November 15th, 2017.
 
+        // The best chain should have at least this much work.
+        consensus.nMinimumChainWork = uint256S("0x0000000000000000000000000000000000000000002cb971dd56d1c583c20f90");
```
Comment (rebroad) [inline src/chainparams.cpp]: Where is this number from please?

Comment (MarcoFalke) [inline src/chainparams.cpp]: @rebroad It is mentioned in the doc change in this very pull...

**Comment by MarcoFalke** [comment]
re-utACK e141beb6a9816b7e1e680fb0a8bae16d42a3e557

**Comment by laanwj** [comment]
utACK e141beb

File: src/chainparams.cpp (hunk 1 above)
Comment (dcousens) [inline src/chainparams.cpp]: @MarcoFalke it mentions the source,  but not the conditions with which to re-create it.
I suppose I would just iterate the blockchain until I found it?

Comment (gmaxwell) [inline src/chainparams.cpp]: You can just run the getblockchaininfo rpc per the instructions and observe that its a number equal to or greater than that in order to verify it. :)

**Comment by sipa** [comment]
utACK e141beb6a9816b7e1e680fb0a8bae16d42a3e557

**Comment by sipa** [comment]
I wonder if this should be backported to 0.13.2, as it fixes a visible issue on testnet.

File: src/chainparams.cpp (hunk 1 above)
Comment (rebroad) [inline src/chainparams.cpp]: @MarcoFalke in the "doc change"? Not sure what you mean.

**Comment by jtimon** [comment]
utACK e141beb6a9816b7e1e680fb0a8bae16d42a3e557

File: src/main.cpp [hunk 2]
```diff
@@ -1740,13 +1740,14 @@ bool IsInitialBlockDownload()
...
+    if (chainActive.Tip() == NULL)
         return true;
-    bool state = (chainActive.Height() < pindexBestHeader->nHeight - 24 * 6 ||
-            std::max(chainActive.Tip()->GetBlockTime(), pindexBestHeader->GetBlockTime()) < GetTime() - nMaxTipAge);
-    if (!state)
-        latchToFalse.store(true, std::memory_order_relaxed);
-    return state;
+    if (chainActive.Tip()->nChainWork < UintToArith256(chainParams.GetConsensus().nMinimumChainWork))
+        return true;
+    if (chainActive.Tip()->GetBlockTime() < (GetTime() - nMaxTipAge))
```
Comment (mruddy) [inline src/main.cpp]: Any particular reason to use GetTime here over GetAdjustedTime?

**Pull request merged**